### PR TITLE
Latex rendering with react-native-math-view

### PR DESCRIPTION
**Pull request by KobeBergmans**
#### Summary

Added rendering of latex code block with the react-native-math-view library. Latex code is split up in different lines and these are rendered independently. By default the code block displays 2 lines of latex code. If you click on the block the next screen will display all the code in a horizontal and vertical scroll view.

To get the inline functionality working I changed the commonmark and commonmark-react-renderer library. The PR's for those library's can be found here:
- https://github.com/mattermost/commonmark.js/pull/11
- https://github.com/mattermost/commonmark-react-renderer/pull/3

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-43469

#### Checklist

- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [ ] Includes text changes and localization file updates

#### Device Information
This PR was tested on: 
- Pixel 5: Android studio emulator
- OnePlus Nord 2 5G: Oxygen OS V11.3
- Iphone 12: Xcode emulator

#### Screenshots

<p float="left">

</p>
<p float="left">

</p>
<p float="left">

</p>

#### Release Note

```release-note
Added rendering of latex code blocks. Added latex screen.
```

### CODE REVIEW

**Review by esethna — COMMENTED** [review]
@KobeBergmans thanks! Does this support the same in-line syntax as the webapp PR you submitted? Similarly, do the config settings we discussed on the webapp PRs also apply here?

**Comment by KobeBergmans** [comment]
@esethna This PR does not include inline rendering of latex it just renders the usual latex code block that was present before my previous PR. I am currently trying to find a solution for the inline latex rendering too but that has not yet worked nicely.

**Comment by KobeBergmans** [comment]
I made the inline latex work like this: 
Inline code

The code does not work 100% correct yet but does this seem like a good enough solution to be implemented?

**Comment by esethna** [comment]
Nice, thanks @KobeBergmans! I'll let @enahum comment on the code/approach.

**Comment by KobeBergmans** [comment]
@enahum @anurag6713 I polished the inline latex functionality and added some tests for it. I merged the code into this PR normally.

**Comment by KobeBergmans** [comment]
@enahum, @anurag6713 I updated the latex rendering to also not trigger when the `$` is preceded by a word character as suggested here: https://github.com/mattermost/mattermost-webapp/pull/8644#issuecomment-918335216

**Comment by KobeBergmans** [comment]
@esethna Is someone still looking in to this? Because I haven't had any feedback from anyone...

File: app/components/markdown/markdown.js [hunk 1]
```diff
@@ -169,6 +179,41 @@ export default class Markdown extends PureComponent {
...
+
+                latexCode = unEscapeLatexCodeInText(latexCode); //Unescape code
+
+                latexCode = latexCode.replace('&', '\\&'); //For some reason the slash before an & character is missing when you type it.
+
+                return (
+                    <Text
+                        style={{lineHeight: this.state.inlineLatexHeight}}
+                    >
+                        {this.renderText({context, literal: firstPart})}
```
Comment (enahum) [inline app/components/markdown/markdown.js:202]: this is calling itself, is that expected?

Also, perhaps you can extract all this if block to it's own function for readability.

File: app/components/markdown/markdown.js [hunk 2]
```diff
@@ -169,6 +179,41 @@ export default class Markdown extends PureComponent {
...
+                        <LatexInline
+                            content={latexCode}
+                            onLayout={(event) => {
```
Comment (enahum) [inline app/components/markdown/markdown.js:205]: do not use inline functions

File: app/screens/latex/latex.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    componentWillUnmount() {
+        BackHandler.removeEventListener('hardwareBackPress', this.handleAndroidBack);
+    }
+
+    handleAndroidBack = () => {
+        popTopScreen();
+        return true;
+    };
```
Comment (enahum) [inline app/screens/latex/latex.js:37]: Any reason why this is needed? is it not handled by default?

File: app/screens/latex/latex.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    render() {
+        const style = getStyleSheet(this.props.theme);
+
+        const lines = splitLatexCodeInLines(this.props.content);
+
+        return (
+            <SafeAreaView
+                edges={['bottom', 'left', 'right']}
```
Comment (enahum) [inline app/screens/latex/latex.js:46]: why is top is being omitted?

File: app/screens/latex/latex.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                                key={latexCode}
+                            >
+                                <MathView
+                                    math={latexCode}
+                                    onError={(errorMsg) => {
+                                        return <Text style={style.errorText}>{'Error: ' + errorMsg.message}</Text>;
+                                    }}
+                                    renderError={(errorMsg) => {
+                                        return <Text style={style.errorText}>{'Render error: ' + errorMsg.error.message}</Text>;
+                                    }}
```
Comment (enahum) [inline app/screens/latex/latex.js:70]: no inline functions

File: app/screens/latex/latex.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        code: {
+            flexDirection: 'row',
+            justifyContent: 'flex-start',
+            marginHorizontal: 5,
+            paddingVertical: codeVerticalPadding,
+        },
+        errorText: {
+            fontSize: 14,
+            marginHorizontal: 5,
+            color: 'rgb(255, 0, 0)',
```
Comment (enahum) [inline app/screens/latex/latex.js:110]: errorTextColor

File: ios/Gemfile.lock [hunk 7]
```diff
@@ -57,6 +57,7 @@ GEM
     ethon (0.12.0)
       ffi (>= 1.3.0)
     ffi (1.15.0)
+    ffi (1.15.0-x64-mingw32)
```
Comment (enahum) [inline ios/Gemfile.lock:60]: why was this added?

File: jest.config.js [hunk 8]
```diff
@@ -28,6 +28,6 @@ module.exports = {
         'assets/images/video_player/(.*).png': '<rootDir>/dist/assets/images/video_player/$1@2x.png',
     },
     transformIgnorePatterns: [
-        'node_modules/(?!(react-native|@react-native|jail-monkey|@sentry/react-native|react-navigation|@react-native-community/cameraroll))',
+        'node_modules/(?!(react-native|@react-native|jail-monkey|@sentry/react-native|react-navigation|@react-native-community/cameraroll|hast-util-from-selector|hastscript|property-information|hast-util-parse-selector|space-separated-tokens|comma-separated-tokens|zwitch))',
```
Comment (enahum) [inline jest.config.js:31]: can you explain why this changes are required ?

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/components/markdown/latex_code_block/latex_code_block.js [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+                                key={latexCode}
+                            >
+                                <MathView
+                                    math={latexCode}
+                                    onError={(errorMsg) => {
+                                        return <Text style={style.errorText}>{'Error: ' + errorMsg.message}</Text>;
+                                    }}
+                                    renderError={(errorMsg) => {
+                                        return <Text style={style.errorText}>{'Render error: ' + errorMsg.error.message}</Text>;
+                                    }}
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply app/components/markdown/latex_code_block/latex_code_block.js:139]: ^ The current code makes it so that the MathView will re-render whenever the LatexCodeBlock does because it's passing a new copy of the arrow function in each time. If they're methods on the LatexCodeBlock, it'll always pass the a reference to the same method.

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: package.json [hunk 10]
```diff
@@ -58,6 +58,7 @@
     "react-native-linear-gradient": "2.5.6",
     "react-native-local-auth": "1.6.0",
     "react-native-localize": "2.1.1",
+    "react-native-math-view": "3.9.4",
```
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply package.json:61]: I started working on this a couple months back and the newest version was not out back then. But i'll see if i can upgrade to the newer version

File: app/components/markdown/latex_code_block/latex_code_block.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply app/components/markdown/latex_code_block/latex_code_block.js:139]: Ah ok i didn't know that. I'll change that to a class method

File: ios/Gemfile.lock (hunk 7 above)
Comment (KobeBergmans) [reply ios/Gemfile.lock:60]: This must have been an accident while trying to get my workspace up and running. I'll delete this

File: jest.config.js (hunk 8 above)
Comment (KobeBergmans) [reply jest.config.js:31]: This happens when a module is not transformed by the typescript compiler. Jest gives following error:

```
    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/en/ecmascript-modules for how to enable it.
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:
    export function zwitch(key, options) {
    ^^^^^^

    SyntaxError: Unexpected token 'export'

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1350:14)
      at Object.<anonymous> (node_modules/hast-util-from-selector/index.js:22:1)
```

So the modules need to be forced to be transformed by jest.

You can read more on this in following thread: https://stackoverflow.com/questions/49263429/jest-gives-an-error-syntaxerror-unexpected-token-export

File: app/components/markdown/latex_code_block/latex_code_block.js [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+import {preventDoubleTap} from '@utils/tap';
+import {changeOpacity, makeStyleSheetFromTheme} from '@utils/theme';
+
+import MarkdownCodeBlock from '../markdown_code_block/markdown_code_block';
+
+const MAX_LINES = 2;
+
+export default class LatexCodeBlock extends MarkdownCodeBlock {
+    static defaultProps = {
+        language: 'latex',
```
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply app/components/markdown/latex_code_block/latex_code_block.js:22]: Seems like a good idea, I just kept it there from the default MarkdownCodeBlock but it is indeed not necessary

File: app/components/markdown/latex_code_block/latex_code_block.js [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+import {splitLatexCodeInLines} from '@utils/latex';
+import {getDisplayNameForLanguage} from '@utils/markdown';
+import {preventDoubleTap} from '@utils/tap';
+import {changeOpacity, makeStyleSheetFromTheme} from '@utils/theme';
+
+import MarkdownCodeBlock from '../markdown_code_block/markdown_code_block';
+
+const MAX_LINES = 2;
+
+export default class LatexCodeBlock extends MarkdownCodeBlock {
```
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply app/components/markdown/latex_code_block/latex_code_block.js:20]: I had a look at changing the files to typescript but it seemed a lot more complex than the webapp because I had no real example to go by. But if I have some time left in the following week I will have a look at that.

In the mean time I'll add the `propTypes`

File: app/components/markdown/latex_code_block/latex_code_block.js [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+        }
+
+        Keyboard.dismiss();
+        requestAnimationFrame(() => {
+            goToScreen(screen, title, passProps);
+        });
+    });
+
+    trimContent = (content) => {
+        const lines = splitLatexCodeInLines(content);
```
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply app/components/markdown/latex_code_block/latex_code_block.js:59]: The library I used to implement the latex code does not support the standard Latex linebreaks (syntax: `\\`). So if this function is not called there are no linebreaks in the rendered code

File: app/components/markdown/markdown.js [hunk 14]
```diff
@@ -446,7 +501,14 @@ export default class Markdown extends PureComponent {
     };
 
     render() {
-        let ast = this.parser.parse(this.props.value.toString());
+        let ast;
+
+        if (this.props.enableLatex && this.props.enableInlineLatex) {
+            const escapedContent = escapeLatexInputPreParser(this.props.value.toString());
```
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply app/components/markdown/markdown.js:507]: The way I solved this in the web app PR was by modifying the markdown parser. I made the parser stop at `$` and then check if there was any latex code ahead. If there was latex code ahead it was parsed into the AST as one big block and the different markdown syntax components where ignored (like `_emphasis_`).

I tried to fix this problem by adding a transformation as you described but this didn't work when there were markdown syntax compontents inside of the inline latex code. For example: `$x_0 = x_1 = /cdots$` would not render correctely because the markdown library would parse `0 = x` as a block that would needed to get emphasis. 

It is possible to overcome this issue by manipulating the AST and checking if there is special syntax inside inline latex code as you said. But that seemed a lot more complicated than doing it this way.

File: app/utils/latex.ts [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+     * These checks are special because they need to check if the braces and statements are not escaped
+     */
+
+    //Check for cases
+    let beginCases = 0;
+    let endCases = 0;
+
+    let i = 0;
+    while (i < latexCode.length) {
+        const firstBegin = latexCode.indexOf('\\begin{', i);
```
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply app/utils/latex.ts:41]: Unfortunaltely the react-native framework is not really that compatible with web library's. I looked into using a webView component to render the latex code but that was an ugly solution because inline this did not work well. So the `katex` library was not really an option on mobile

The `react-native-math-view library` is a way to fix this issue, but the drawback is that it does not support line breaks. So they need to be added manually

File: app/utils/latex.ts [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    let i = 0;
+    while (i < outLines.length) {
+        if (testLatexLineBreak(outLines[i])) { //Line has no linebreak in between brackets
+            i += 1;
+        } else if (i < outLines.length - 2) {
+            outLines = outLines.slice(0, i).concat([outLines[i] + '\\\\' + outLines[i + 1]], outLines.slice(i + 2));
+        } else if (i === outLines.length - 2) {
+            outLines = outLines.slice(0, i).concat([outLines[i] + '\\\\' + outLines[i + 1]]);
+        } else {
+            return outLines;
```
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply app/utils/latex.ts:23]: That is indeed true, I missed that

File: app/utils/latex.ts [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+// See LICENSE.txt for license information.
+
+/**
+ * Splits up latex code in an array consisting of each line of latex code.
+ * A latex linebreak is denoted by `\\`.
+ * A line is not broken in 2 cases:
+ * - The linebreak is in between brackets.
+ * - The linebreak occurs inbetween a `\begin` and `\end` statement.
+ */
+export function splitLatexCodeInLines(content: string): Array<string> {
```
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply app/utils/latex.ts:11]: The linebreaks inside of a begin/end statement or curly braces are handled correctely by the `react-native-math-view` library.

For example when creating a matrix:
```
\begin{matrix}
1 & 2 & 3\\
a & b & c
\end{matrix}
```

The matrix still needs to be printed on 1 line because the linebreaks appear inside of the matrix statement. 

The same thing happens when a line break is in between curly braces. An example of this is the following:
```
\prod_{\substack{1\le i \le n\\ 1\le j \le m}}M_{i,j}
```

File: app/components/markdown/markdown.js (hunk 1 above)
Comment (KobeBergmans) [reply app/components/markdown/markdown.js:202]: That behaviour is expected. What this function does is check if there is an inline latex match. If there is a match it splits the text up in text and latex code. It calls itself on the text part to check if there are maybe more inline latex code blocks and renders the latex code.

I'll split this function up.

File: app/screens/latex/latex.js (hunk 3 above)
Comment (KobeBergmans) [reply app/screens/latex/latex.js:37]: I copied this from the code screen but i think it might be a better solution to just extend the Latex screen from the code screen. I'll implement that so there is no duplicate code

File: app/screens/latex/latex.js (hunk 4 above)
Comment (KobeBergmans) [reply app/screens/latex/latex.js:58]: I don't know why the top is omitted but i based my code on the `Code` screen and there it is omitted too.

File: app/utils/latex.ts [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    //Check for braces
+    let curlyOpenCases = 0;
+    let curlyCloseCases = 0;
+
+    i = 0;
+    while (i < latexCode.length) {
+        const firstBegin = latexCode.indexOf('{', i);
+        const firstEnd = latexCode.indexOf('}', i);
```
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply app/utils/latex.ts:72]: I don't know if you wanted to have the double backslash or single so i did both.

The single backslash `\begin{}}\end{}` is not split up.
Double backslash `\\begin{}}\\end{}` is split in 2 lines. A first empty line and a second one that has `begin{}}\\end{}` wich gives a render error

**Comment by KobeBergmans** [comment]
@enahum, @hmhealey I think I got to all the requested changes, if I missed one or you have more questions let me know!

File: app/components/markdown/latex_code_block/latex_code_block.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply app/components/markdown/latex_code_block/latex_code_block.js:21]: Works for me. It's been a while since I've worked on the mobile app, and I know most of the TypeScript migration effort has been on the web-side, so I'm not surprised that it's trickier to do here.

File: app/components/markdown/latex_code_block/latex_code_block.js [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+
+    handlePress = preventDoubleTap(() => {
+        const {language, content} = this.props;
+        const {intl} = this.context;
+        const screen = 'Latex';
+        const passProps = {
+            content,
+        };
+
+        const languageDisplayName = getDisplayNameForLanguage(language);
```
Comment (hmhealey) [inline app/components/markdown/latex_code_block/latex_code_block.js:36]: To follow up on https://github.com/mattermost/mattermost-mobile/pull/5651#discussion_r718550915, could you remove the language prop from here as well?

File: app/utils/latex.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply app/utils/latex.ts:41]: I see what you mean. This feels like why I ended up writing so much of the mobile app's Markdown code from scratch. Unfortunately, react-native has a lot fewer libraries for things like this, so we end up having to do a lot ourselves.

File: app/components/markdown/markdown.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply app/components/markdown/markdown.js:522]: Perhaps it would be easier to just modify the Markdown library to do that then. I think I'd feel more comfortable doing that than escaping and unescaping these since the current way of doing this feels like it could be fragile since anything else handled to the Markdown code could interfere with this as it currently is. For example, if at-signs or number signs are used in the Latex, I think that'll cause the Markdown parser to interfere since the mobile app's version is actually responsible for doing at-mentions and hashtags as well.

Overall, I don't think it should be that hard to add since [our fork of commonmark.js](https://github.com/mattermost/commonmark.js) is much better maintained, and I've generally found it easier to work with than marked, but it does require changes to two libraries to change: that fork and https://github.com/mattermost/commonmark-react-renderer.

You'll have to make most of your changes in [`inlines.js`](https://github.com/mattermost/commonmark.js/blob/master/lib/inlines.js) by adding an extra case to `parseInlines`. It's a bit more powerful than the `marked` version, so you can backtrack to check what's before the $ without having to use a lookbehind (see the at mention or hashtag parsing for an example of that).

Then, you'll want to add a new `Node` type for the inline latex and ensure commonmark-react-renderer has some basic handling for it since it needs to know about every potential type of element for some reason that I can't quite remember.

Apologies for making this more complicated, but I think avoiding the escaping logic will be safer in the long run.

File: app/utils/latex.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply app/utils/latex.ts:11]: Ah, I see. So that's needed because the react-native-math-view doesn't know how to handle the line breaks itself when it would better understand how to apply them in different contexts.

Thanks for the examples since I've only got a very limited understanding of Latex.

File: app/utils/latex.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply app/utils/latex.ts:72]: Cool. As long as it errors nicely without crashing the app, that's good with me. Taking another look at this, I can tell that it won't ever loop infinitely, and that's what I was most worried about anyway.

**Review by esethna — APPROVED** [review]

File: app/components/markdown/markdown.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (KobeBergmans) [reply app/components/markdown/markdown.js:522]: I understand that this way of doing things is indeed a bit safer and more robust so I'll take a look at this in the following week.

I thought about changing the commonmark library when I started working on this PR but I looked at the packages and it seemed like the app used the standard commonmark packages. Then I looked into the patches that are done after the npm install process and I guess that is where the changes occur.

So how do I change the commonmark library's? Do I just make a PR on the mattermost fork and you guys will change the patch file or how does this work?

**Comment by KobeBergmans** [comment]
@enahum @hmhealey I played around with the patch files. I think I did everything right but on my machine the patch does not work, the inline latex features are broken. Can someone maybe generate the right files for me? Because I can't seem to find the mistake.

**Comment by hmhealey** [comment]
@enahum Do we have any docs for doing that, or are you able to help generate those?

**Comment by enahum** [comment]
it seems like the fork is behind our commonmark fork, perhaps that is the problem

image

image

**Comment by KobeBergmans** [comment]
@enahum Thanks for the tip I made the patches from the wrong repository. 

Now everything works as espected with the patches.

**Review by hmhealey — APPROVED** [review]
Cool, the patches look good to me

**Review by enahum — APPROVED** [review]
this LGTM. @Willyfrog can you create a ticket for this to get ported to Gekidou? Thanks!!

**Comment by enahum** [comment]
@KobeBergmans can you please resolve the conflicts ?

**Comment by Willyfrog** [comment]
> this LGTM. @Willyfrog can you create a ticket for this to get ported to Gekidou? Thanks!!

[done](https://mattermost.atlassian.net/browse/MM-40194)

**Comment by KobeBergmans** [comment]
I think i resolved all the conflicts by merging the master branch @enahum

**Review by anurag6713 — APPROVED** [review]

**Review by josephbaylon — CHANGES_REQUESTED** [review]
@KobeBergmans Here's the result of initial testing for both iOS and Android. Bolded ones are potential issues. Please look into.

A. Latex code preview and opening latex code itself
```
\```latex
X_k = \sum_{n=0}^{2N-1} x_n \cos \left[\frac{\pi}{N} \left(n+\frac{1}{2}+\frac{N}{2}\right) \left(k+\frac{1}{2}\right) \right]
\```
```
1. Works fine on Android
2. Works fine on iOS

B. Inline latex
```
a couple of equations in 1 message: $6^2 - 5 = 36-5 = 31$, another one: $L' = {L}{\sqrt{1-\frac{v^2}{c^2}}}$. Maxwell: $E'=\nabla \times B - 4\pi j$. A hard equation: $\lim_{x\to 0}{\frac{e^x-1}{2x}} \overset{\left[\frac{0}{0}\right]}{\underset{\mathrm{H}}{=}} \lim_{x\to 0}{\frac{e^x}{2}}={\frac{1}{2}}$. Some weird symbols: $y = a + f(\underbrace{b x}_{\ge 0 \text{ by assumption}}) = a + f(\underbrace{b x}_{\mathclap{\ge 0 \text{ by assumption}}})$. Verry long arrows: $A \xleftarrow{\text{this way}} B \xrightarrow[\text{or that way}]{ } C$. Sample integral: $b_n=\frac{1}{\pi}\int\limits_{-\pi}^{\pi}f(x)\sin nx\,\mathrm{d}x=\frac{1}{\pi}\int\limits_{-\pi}^{\pi}x^2\sin nx\,\mathrm{d}x$
```
1. Works fine on Android
2. **Overlaps on iOS (See image)**
ios_latex_inline

C. Inline latex triggers
```
1. some text with 15$ and 20$
2. some text with $15 and $20
3. some text with 15$ and $20
4. some text with $20 and 20$
5. [$1+2=3$](https://example.com)
6. _$1+2=3$_
7. **$1+2=3$**
8. ~~$1+2=3$~~
9. https://example.com/$/test/$foo
10.
\```
$1+2=3$
\```
11. `$1+2=3$`
```
1. **On Android, items 5 and 6 do not seem to render properly (See image)**
android_latex_inline_trigger

2. **On iOS, items 5, 6, 7, 8 do not seem to render properly (See image)**
ios_latex_inline_trigger

D. Inline rainbow
```
Inline-rainbow
# $\color{red}r\color{orange}a\color{yellow}i\color{green}n\color{blue}b\color{Indigo}o\color{violet}w$
```
1. Works fine on Android
2. **Blank on iOS**

**Comment by KobeBergmans** [comment]
oh I totaly forgot to take a look at this... I'll take a look at it right now.

**Comment by KobeBergmans** [comment]
I took a look at the initial testing you did @josephbaylon. I looked into the Android problems in particular because at this time I don't have acces to a mac to test the IOS bugs, I did try to comment on them to help you in further testing. If there is a way I can test and resolve these issues without a mac please point them out and I'll take a look at it.

B, Iphone:
The overlap did happen to me too 1 or 2 times if I was testing on the android emulator. I never got this on my android device though so I figured it was due to the emulator running slow. Did you test this on a real device and did it always happen?

C, 5,6 (both platforms):
I tried to figure out why this was not rendering properly but i could not find out why... It must be that the commonmark library works differently than the marked library (used for the webapp).
I did find out that the problem only occurs for links and emphasis when the `_` symbol is used (with `*` it works fine). So I guess it has to do with `_` and `[]` because it works for the other delimiters? Maybe @hmhealey can help me on this one?

C, 7,8, Iphone:
I fear that this has to do with the react-native-math-view library... I can't find any other reason for the different behaviour on both platforms. So I think the library can't render latex properly on IOS if it is in an environement where there are certain styles applied.
To test this further can you maybe test is `*$1+2$*` renders?

**Comment by josephbaylon** [comment]
@KobeBergmans You can maybe use a mac virtual machine to test on iOS simulator.

B, Iphone: I test on real device (iPhone X 15.2) and the overlap always happens
C, 7,8, Iphone: `*$1+2$*` does not render on iPhone but renders on Android

I'll let @hmhealey and @enahum chime in on `C, 5,6 (both platforms)` and `D, Iphone`
https://github.com/mattermost/mattermost-mobile/pull/5651#issuecomment-985403167
https://github.com/mattermost/mattermost-mobile/pull/5651#pullrequestreview-812937264

**Comment by enahum** [comment]
C. 5 seems to be a link and I would expect it to render as such
C. 6 to 8 I would expect to render the text as italic, bold, strikethrough on both platforms
D is probably not supported by the new library on iOS (not sure if that is the actual case)

So for me C5 is fine, C6-8 regular markdown should win and render properly and D needs confirmation, as if that is the case we can add it as a limitation in the release notes.

@hmhealey thoughts on this?

**Comment by hmhealey** [comment]
Yeah, I agree that C5-8 should prioritize the Markdown over Latex. I think the reason C7 behaves differently is because of [this line](https://github.com/mattermost/commonmark.js/blob/master/lib/inlines.js#L1147) since `_` is a word character in regex. To make it not parse inline latex inside of a link or bold/italic delimiters, I think you can check if `this.brackets` and `this.delimiters` respectively, and if they're both falsey, you should be safe to check for inline latex.

For D, I agree that that one could be a known issue since I think Elias is right that it's a library issue

**Comment by KobeBergmans** [comment]
Ok I'll try to take a look at this during this weekend. I'll try to have a go at the mac emulator.

Is it necessary that the webapp and the mobile app have the same functionality because I think `C5-8` do render as inline latex in the webapp so maybe that needs some changing too @hmhealey?

I also came up with the same conclusion on the word character but I changed the regex to not include word characters and it did not work. But I'll take another look at that this weekend.

**Comment by hmhealey** [comment]
I think that's fine if they behave differently in those cases even if it's not 100% ideal. Eventually, both the mobile and web app will both be using commonmark.js, so they should behave the same then anyway.

**Comment by KobeBergmans** [comment]
> Yeah, I agree that C5-8 should prioritize the Markdown over Latex. I think the reason C7 behaves differently is because of [this line](https://github.com/mattermost/commonmark.js/blob/master/lib/inlines.js#L1147) since `_` is a word character in regex. To make it not parse inline latex inside of a link or bold/italic delimiters, I think you can check if `this.brackets` and `this.delimiters` respectively, and if they're both falsey, you should be safe to check for inline latex.
> 
> For D, I agree that that one could be a known issue since I think Elias is right that it's a library issue

This solution seemed to work for me so I implemented it and added another PR for the commonmark library. so problem C is now solved ;)

I am still working on the mac VM for problem B

**Comment by KobeBergmans** [comment]
I finally got the MacOS VM working on my machine, I'll try to get a look at the problem in the folowing week

**Comment by KobeBergmans** [comment]
> @KobeBergmans any progress on this?

I am trying to get everything working on the mac vm but it is not that easy it seems. I currently also have exams so I don't have a lot of time to look into this...

**Comment by esethna** [comment]
Hey @KobeBergmans, just checking in on this. Please let us know if you have questions or are still working on this.

**Comment by KobeBergmans** [comment]
> Hey @KobeBergmans, just checking in on this. Please let us know if you have questions or are still working on this.

Hey @esethna I am still trying to get things working on my Mac VM  to solve the problem on Iphone but I think I am getting there. It is pretty busy for me these days though thus I think it will take some time before it gets resolved

**Review by josephbaylon — APPROVED** [review]
LGTM. Latex rendering is now working properly on both iOS and Android. Previous issues found are now fixed. Thanks @KobeBergmans !!

**Comment by enahum** [comment]
Created a ticket to port this functionality to v2 https://mattermost.atlassian.net/browse/MM-43510

**Comment by KobeBergmans** [comment]
No problem @josephbaylon I am terribly sorry for the long wait it should not have taken this long...

**Comment by enahum** [comment]
I'm going to manually cherry pick this one as well.
